### PR TITLE
SONARPY-2110 Make the ProjectLevelSymbolTable return the Descriptor model instead of the v1 Symbol model

### PR DESCRIPTION
**Pull request by ghislainpiot**
This pull request introduces several changes to the `python-frontend` module, focusing on enhancing the handling of Python type descriptors and improving type conversion mechanisms. The most important changes include the addition of a new method to retrieve descriptors from a module, updates to the `SymbolsModuleTypeProvider` class to use different type converters, and modifications to the `AnyDescriptorToPythonTypeConverter` class to handle type origins.

### Enhancements to Type Descriptor Handling:

* [`python-frontend/src/main/java/org/sonar/python/semantic/ProjectLevelSymbolTable.java`](diffhunk://#diff-b7ec63b867f8e4e30a2a2b2bc9d6f30cc63a272de1ec7d2d6795323270764a18R167-R171): Added a new method `getDescriptorsFromModule` to retrieve descriptors from a module.

### Updates to SymbolsModuleTypeProvider:

* [`python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java`](diffhunk://#diff-768ca79d1818e5bdd6d0645830cbe0a58fb72767a54437eef94bf17face10370L22-R50): Updated to use `stubsDescriptorToPythonTypeConverter` and `projectDescriptorToPythonTypeConverter` instead of a single `anyDescriptorToPythonTypeConverter`. This change ensures that type origins are correctly handled. [[1]](diffhunk://#diff-768ca79d1818e5bdd6d0645830cbe0a58fb72767a54437eef94bf17face10370L22-R50) [[2]](diffhunk://#diff-768ca79d1818e5bdd6d0645830cbe0a58fb72767a54437eef94bf17face10370L91-L233)

### Modifications to Type Converters:

* [`python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/AnyDescriptorToPythonTypeConverter.java`](diffhunk://#diff-9ab2859150eb1d2ce8046c80a8f06907c18654823452a9dd1a33d040e9c48b41R27-R49): Enhanced to include `TypeOrigin` as a parameter, allowing the differentiation between stub and local types during conversion.
* [`python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/ConversionContext.java`](diffhunk://#diff-cbf8d9b5d10e51dabc8d226cf593eab2a19aff001cabb41150993634f2711b52R27-R49): Updated to store and utilize `TypeOrigin` during type conversion processes.

### Test Adjustments:

* Various test files (`ConversionContextTest`, `DescriptorToPythonTypeConverterTest`, `TypeInferenceV2Test`) were updated to reflect the changes in type conversion and ensure the correctness of the new implementations. [[1]](diffhunk://#diff-c19b02e07e0ce0cfe7e25bfee028c73cc1d2b8d11d9820f485bd7f0b139c258cL725-R725) [[2]](diffhunk://#diff-2dc101bd60c21461535f12c275d062d8d256d97e26b98ba3d8af6c1098b3eac8R28-R36) [[3]](diffhunk://#diff-33e5c389e0cbac00af5f958a77f0456916c6aa9978c0dfeb23aa622016888b91R44)

### CODE REVIEW

**Review by guillaume-dequenne-sonarsource — COMMENTED** [review]
Looks good, a few minor questions.

File: python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java [hunk 1]
```diff
@@ -722,7 +722,7 @@ void inferFunctionParameterTypesMultiFile() {
     assertThat(fooType.parameters().get(0).declaredType().type().unwrappedType()).isEqualTo(intType);
 
     FunctionType foo2Type = (FunctionType) ((ExpressionStatement) fileInput.statements().statements().get(2)).expressions().get(0).typeV2();
-    assertThat(foo2Type.parameters()).extracting(ParameterV2::declaredType).extracting(TypeWrapper::type).containsExactly(dictType, aType);
+    assertThat(foo2Type.parameters()).extracting(ParameterV2::declaredType).extracting(TypeWrapper::type).extracting(PythonType::unwrappedType).containsExactly(dictType, aType);
```
[the remark this answers is not in the document]
Comment (ghislainpiot) [reply python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java:725]: That is correct, now everything is wrapped in ObjectType for the parameters.

Also, we now have a TypeWrapper of an ObjectType containing a TypeWrapper for each parameter. If we decide to keep the ObjectType wrapping, maybe we could drop the outer TypeWrapper and just have a PythonType ?

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java [hunk 2]
```diff
@@ -88,147 +68,20 @@ private static String getModuleFqnString(List<String> moduleFqn) {
   }
 
   private Optional<ModuleType> createModuleTypeFromProjectLevelSymbolTable(String moduleName, String moduleFqn, ModuleType parent) {
-    return Optional.ofNullable(projectLevelSymbolTable.getSymbolsFromModule(moduleFqn))
-      .map(projectModuleSymbols -> createModuleFromSymbols(moduleName, parent, projectModuleSymbols));
+    var retrieved = projectLevelSymbolTable.getDescriptorsFromModule(moduleFqn);
+    if (retrieved == null) {
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:72]: can't it be converted just to `Optional.ofNullable(...)...`, then you simply don't need this condition like:
```
return Optional.ofNullable(projectLevelSymbolTable.getDescriptorsFromModule(moduleFqn))
 .map(retrieved -> retrieved.stream().collect(Collectors.toMap(Descriptor::name, d -> anyDescriptorToPythonTypeConverter.convert(d, TypeOrigin.LOCAL))))
 .map(members -> new ModuleType(moduleName, parent, members))
```

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**